### ISSUE TITLE
Cross-validation in the problem set

### ISSUE THREAD

**Issue by FranciscoRMendes** [macss-modeling/General-Questions#8]
Dear Professor

I have a few questions on the problem set 2. 

1. What does it mean to "fit" a logistic regression using 10-fold cross validation? As there are no hyper parameters what exactly is being "fit" here? I get that it is useful for checking how well a model performs but it seems like the prompt asks us to evaluate the model based on the test set. 
2. How exactly do we choose the final model? If we evaluate the model using the test set (As per prompt), R will simply take the entire train set and fit a model and predict (this defeats the purpose of 1). 
3. How do we generate an error rate before choosing a threshold from the ROC?

Thanks
Franco

**Comment by pdwaggoner** [comment]
Hi - Thanks for the questions. 

1. Fit a logistic regression or any model for that matter, with or without hyperparameters, using cross validation is how we test the generalizability of the models we train. That is, recall the goal from the beginning of the class: build generalizable, good models. Because we can't draw repeated samples from the same population, we have to emulate the process with the sample we have. Thus, we use CV to do so and get a measure of so-called generalization error. 

2. Almost, but you're misreading the question; it doesn't defeat the purpose of the use of CV to generate generalization/test error. The final model, comparing logit to all versions of kNN, LDA, and QDA, *not* comparing multiple version of logit only. So fit the models, compared their error in a variety of ways mentioned in the problem set, and then pick the one with the lowest CV error. And continue. 

3. Not sure what you're asking, but these are separate measures. `err = 1 - accuracy`, the ROC curve is a plot version of the AUC, whereas AUC is bounded between 0 and 1, with 1 being maximal area under the ROC curve, signifying the best fitting model, and 0 == random guess/the 45 degree line in an ROC plot. 

I hope this helps.

**Comment by FranciscoRMendes** [comment]
Dear Professor

Thank you for your answer!
Just so that I am 100% sure. 
So when picking a final model, how exactly should we do it?
Suppose we had only QDA, LDA and Logit. 
Should we compare across 10 folds each of the QDA, LDA and Logit? That gives a total of 10 x 3 = 30?
Or should we just compare performance of the final model on the test set?

For the second part, what threshold do I pick to generate an accuracy, if the output from the above models is a probability?

**Comment by pdwaggoner** [comment]
I would trust the default accuracy (0.5). 

And for picking the best model, recall you are to pick based on the CV procedure, meaning you will b e comparing the *cross-validation error rate*, which is your measure of generalization error, across all models. So in your scenario, it wouldn't be 30, it would be 3, because remember the final step in CV, is to average by $k$, which in your scenario is 10. From there, the "best" is however you want to define it. Personally, I would hope the measure of accuracy would agree across the classifiers, but they may not. I would just pick one and make a case for it. E.g., suppose knn with k =7 had the lowest error rate (thatis, misclassifications). I'd pick that as the "best" as it made the fewest mistakes compared to this other. This has the advantage of being a more interpretable measure than AUC, e.g., Just my two cents. I hope this helps!

**Comment by FranciscoRMendes** [comment]
thanks again for your detailed answers!
I am clear on both my questions now.

**Comment by pdwaggoner** [comment]
Absolutely! happy to help

**Comment by Qiuyu-Li** [comment]
Sorry but I think I still don't understand what should we do for part 2 question e. In previous exercises, we fit one model with training data and apply it to the test data. But here, the 10-fold cross-validation provides us with 10 models. Shall we pick the one with the least error rate as the final fitted model from question d, for applying the predict() function to calculate the test error rate? If not, how do we exactly "Evaluate each model’s performance using the test set"? Thank you!

**Comment by pdwaggoner** [comment]
@Qiuyu-Li No, CV doesn't just provide 10 models. It provides 10 models *and then a cv error rate*, which is the summed error across all 10 versions of the model ,averaged by $k$. This is the proxy for generalization/test error as stated many times in a few places. If still confused, please go back through replies on this repo, as well as back to the notes in week 2 on bias, variance, optimism in training error, cross validation, etc. Thanks.

**Comment by Qiuyu-Li** [comment]
Hi Professor, thank you so much for your reply and I think the point that I missed is that: I only read question e but not f... in my previous understanding, "test set" uniquely refers to the set that we use for testing the final model, after tunning hyperparameters with CV. But after I read question f, I realized that the "test set" in e actually refers to "validation set".

Now I understand the question and thank you for your time!

**Comment by pdwaggoner** [comment]
Absolutely!
